ADR/0015-integration-service-two-phase-architecture.md

```diff
@@ -0,0 +1,94 @@
+# 15. The Two-phase Architecture of the Integration Service
+
+Date Documented: 2023-02-09
+
+## Status
+
+In review
+
+## Context
+
+The Integration Service is in charge of running integration test pipelines by executing
+the Tekton pipeline for each user-defined IntegrationTestScenario.
+The main payload that’s being tested are Snapshots which contain references to all
+Components that belong to the Application, along with their images.
+
+A problem we face: Since testing an application with Tekton pipelines takes time,
+and multiple component builds can happen in quick succession, there is potential
+for race conditions between different Snapshots that are created for each of those builds.
+This would primarily manifest itself by two Snapshots getting promoted in quick
+succession, but neither of them having the latest images contained within them.
+
+## Decision
+
+In order to tackle the race conditions, a two phase approach to integration testing
+will take place. The phases will consist of a Component phase that will always be executed
+and the optional Composite phase which would come into play only when race conditions
+between component builds are detected.
+
+### Component phase
+
+When a single component image is built, the Integration Service tests the application
+by creating a Snapshot. All components from the Global Candidate List are included
+and the Component that was built is updated/overwritten with the newly built image.
+
+![](../diagrams/ADR-0015/component-phase.jpg)
+
+### Composite phase
+
+When multiple components are built at the same time, the Integration Service tests
+the application by creating a composite Snapshot using multiple components updated
+to use the newly built images.
```

Comment:
Can you provide details for how this composite Snapshot is created/updated as more components are built?  Here are some questions/self-answers which I think would be good to address/correct.

It doesn't have to be too detailed, but maybe some form of text description about how the Global Candidate List is created/maintained/checked.

Based on later wording, I assume that the global composite list is updated after the first phase by updating a single component. If the Global Candidate List is identical to the just-passed snapshot, then we are all good. If not, the second loop is run. Is this Global Candidate List what is used for sending snapshot references to other services or would a new Snapshot (the composite snapshot?) be created based off of the (then) Global Candidate List (Is it any different from the other Snapshot other than it is built off of the Global Candidate List)? 

It looks like Snapshots will also only be completely tested/promoted if it matches the Global Candidate List, correct?

Reply:
I updated the first phase to clarify that the Global Candidate List is updated only after the testing successfully finishes and the Global Candidate List hasn't changed in the meantime. I also added
```
The Composite phase is used when the Global Candidate List changes while testing a
Snapshot in the Component phase. The Integration service checks this before promoting
a Snapshot at the end of the Component phase.
```

to the description of the Composite phase. Let me know if that clears it up.

Reply:
I think that does clear many things up. So in these situations, the Snapshot which was out of sync with the Global Candidate List will just be "discarded" in favor of the new composite snapshot created from the Global Candidate List? 

I know that this isn't the realm of the integration service, but how would this be detected/displayed in the UI, or is there no promotion representation in the UI yet?

Reply:
Yes, that's correct, we mark the discarded Snapshot's status as invalid with the reason that it was superseded by a composite snapshot - this could be picked up by the UI in the future.

Reply:
Cool. I am good to resolve this thread after we indicate that a new composite snapshot is created based off of the global candidate list and that the previously tested snapshot is discarded by being marked with an invalid status.